### PR TITLE
CLOUDP-78662: Use new readiness probe

### PR DESCRIPTION
**Pull request by chatton**
### All Submissions:

* [N/A] Have you opened an Issue before filing this PR?
* [X] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [X] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).

~~This should be reviewed after: #347~~

This PR updates the operator to use the new readiness probe (reading agent status from the pods themselves)

There were a few additional steps that needed to be taken to get everything working.

* The probe is now provided via an init container and the StatefulSet spec is updated accordingly.
* Now that we have deterministic checking of the agent status, we no longer need to spread partial TLS upgrade logic in different parts of the reconcile loop, we can now do this all in one place, with the only thing that needs to change being the order of the AutomationConfig update vs the StatefulSet update. These two updates have been moved to one place, with the order of operations depending on the state of the resource.
* Removed code around scale down detection, as the new readiness probe handles this for us.

### CODE REVIEW

File: controllers/replica_set_controller.go [hunk 1]
```diff
@@ -182,22 +178,6 @@ func (r ReplicaSetReconciler) Reconcile(ctx context.Context, request reconcile.R
 		)
 	}
 
-	// if we're scaling down, we need to wait until the StatefulSet is at the
-	// desired number of replicas. Scaling down has to happen one member at a time
-	if scale.IsScalingDown(mdb) {
```
Comment (chatton) [inline controllers/replica_set_controller.go:187]: the check for scaling down no longer needs to happen as the readiness check later on accepts the expected number of replicas that should be ready.

File: controllers/replica_set_controller.go [hunk 2]
```diff
@@ -331,60 +289,103 @@ func (r ReplicaSetReconciler) Reconcile(ctx context.Context, request reconcile.R
...
+
+	currentSts, err := r.client.GetStatefulSet(mdb.NamespacedName())
+	if err != nil {
+		return false, fmt.Errorf("error getting StatefulSet: %s", err)
+	}
+
+	r.log.Debugf("Ensuring StatefulSet is ready, with type: %s", getUpdateStrategyType(mdb))
+
+	isReady := statefulset.IsReady(currentSts, mdb.StatefulSetReplicasThisReconciliation())
+	return isReady || currentSts.Spec.UpdateStrategy.Type == appsv1.OnDeleteStatefulSetStrategyType, nil
```
Comment (chatton) [inline controllers/replica_set_controller.go:308]: we check for the OnDelete strategy type here as we want to be considered ready in the case of a version change even if the sts itself is not ready.

File: controllers/replicaset_scaledown.go [hunk 3]
```diff
@@ -21,38 +16,3 @@ func updateScalingStatus(statusWriter k8sClient.StatusWriter, mdb mdbv1.MongoDBC
 	)
 	return err
 }
-
-// checkIfStatefulSetMembersHaveBeenRemovedFromTheAutomationConfig ensures that the expected number of StatefulSet
-// replicas are ready. When a member has its process removed from the Automation Config, the pod will eventually
-// become unready. We use this information to determine if we are safe to continue the reconciliation process.
-func checkIfStatefulSetMembersHaveBeenRemovedFromTheAutomationConfig(stsGetter statefulset.Getter, statusWriter k8sClient.StatusWriter, mdb mdbv1.MongoDBCommunity) (reconcile.Result, error) {
```
Comment (chatton) [inline controllers/replicaset_scaledown.go:28]: existed as part of the scale down check - can be removed completely

File: scripts/dev/e2e.py [hunk 4]
```diff
@@ -164,7 +164,8 @@ def create_test_pod(args: argparse.Namespace, dev_config: DevConfig) -> None:
                         "go",
                         "test",
                         "-v",
-                        "-timeout=60m",
+                        "-timeout=30m",
+                        "-failfast",
```
Comment (chatton) [inline scripts/dev/e2e.py:168]: this was added to prevent the tests from continuing when they had already failed.

File: scripts/dev/e2e.py [hunk 5]
```diff
@@ -164,7 +164,8 @@ def create_test_pod(args: argparse.Namespace, dev_config: DevConfig) -> None:
                         "go",
                         "test",
                         "-v",
-                        "-timeout=60m",
+                        "-timeout=30m",
```
Comment (chatton) [inline scripts/dev/e2e.py:167]: the new readiness probe prevents the need to wait for so long.

File: controllers/replica_set_controller.go [hunk 6]
```diff
@@ -77,11 +82,6 @@ const (
 	// configured
 	lastVersionAnnotationKey    = "mongodb.com/v1.lastVersion"
 	lastSuccessfulConfiguration = "mongodb.com/v1.lastSuccessfulConfiguration"
-	// tlsRolledOutAnnotationKey indicates if TLS has been fully rolled out
-	tlsRolledOutAnnotationKey      = "mongodb.com/v1.tlsRolledOut"
-	hasLeftReadyStateAnnotationKey = "mongodb.com/v1.hasLeftReadyStateAnnotationKey"
-
-	trueAnnotation = "true"
```
Comment (chatton) [inline controllers/replica_set_controller.go:84]: tls annotation checking is no longer required, this is now accounted for when determining the order of the sts/automation config updates.

File: controllers/replica_set_controller.go [hunk 7]
```diff
@@ -519,37 +502,37 @@ func getCustomRolesModification(mdb mdbv1.MongoDBCommunity) (automationconfig.Mo
 	}, nil
 }
 
-func (r ReplicaSetReconciler) buildAutomationConfigSecret(mdb mdbv1.MongoDBCommunity) (corev1.Secret, error) {
+func (r ReplicaSetReconciler) buildAutomationConfig(mdb mdbv1.MongoDBCommunity) (automationconfig.AutomationConfig, error) {
```
Comment (chatton) [inline controllers/replica_set_controller.go:505]: extracted this function into two separate ones, one that builds the Automation Config and one that builds the secret.

File: pkg/util/functions/functions.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+		return runInOrder(funcs...)
+	}
+	return runReversed(funcs...)
+}
+
+func runInOrder(funcs ...func() (bool, error)) (bool, error) {
+	for _, fn := range funcs {
+		ready, err := fn()
+		if err != nil {
+			return ready, err
```
Comment (bznein) [inline pkg/util/functions/functions.go:18]: [micro-nit] Can we find a better name than `ready`? This package is quite generic and `ready` seems to be more suited for a specific case.

File: pkg/util/functions/functions.go [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+// RunSequentially executes a series of functions sequentially. Each function returns a boolean
+// indicating if the function was successfully, and an error indicating if something went wrong.
+// if any function returns an error, an early exit happens. The first parameter indicates if the functions
+// should run in the order provided. A value of true indicates they should run in reverse.
+func RunSequentially(reversed bool, funcs ...func() (bool, error)) (bool, error) {
+	if !reversed {
+		return runInOrder(funcs...)
+	}
+	return runReversed(funcs...)
```
Comment (bznein) [inline pkg/util/functions/functions.go:11]: [micro-nit] I find it more logic to swap the two, but it's 100% up to you :D

File: pkg/util/functions/functions.go [hunk 10]
```diff
@@ -0,0 +1,38 @@
+package functions
+
+// RunSequentially executes a series of functions sequentially. Each function returns a boolean
+// indicating if the function was successfully, and an error indicating if something went wrong.
+// if any function returns an error, an early exit happens. The first parameter indicates if the functions
+// should run in the order provided. A value of true indicates they should run in reverse.
+func RunSequentially(reversed bool, funcs ...func() (bool, error)) (bool, error) {
```
Comment (bznein) [inline pkg/util/functions/functions.go:7]: [q] why are we using a boolean that tells us if we want to flip the order of execution?

I'm thinking of how a dev would interpret seeing this function called:

`RunSequentially(true, functions...)`

would make me thing that these will be ran in order, not in the opposite one.

File: test/e2e/mongodbtests/mongodbtests.go [hunk 11]
```diff
@@ -160,7 +160,7 @@ func CreateMongoDBResource(mdb *mdbv1.MongoDBCommunity, ctx *e2eutil.Context) fu
 
 func BasicFunctionality(mdb *mdbv1.MongoDBCommunity) func(*testing.T) {
 	return func(t *testing.T) {
-		t.Run("Config Map Was Correctly Created", AutomationConfigSecretExists(mdb))
+		t.Run("Secret Was Correctly Created", AutomationConfigSecretExists(mdb))
```
Comment (bznein) [inline test/e2e/mongodbtests/mongodbtests.go:163]: 👍🏻

**Review by bznein — COMMENTED** [review]
Looking really good! 
I think functionally this is ready to be merged, I just have a couple of comments about clarity of the code in one particular file

File: pkg/util/functions/functions.go (hunk 10 above)
Comment (chatton) [reply pkg/util/functions/functions.go:7]: ah, yes that probably makes more sense! I'll do that refactor 👍

**Comment by chatton** [comment]
@bznein ready for another look when you have time!

**Review by bznein — APPROVED** [review]
LGTM!

**Pull request merged**